server/src/main/java/org/opensearch/plugins/EnginePlugin.java

```diff
@@ -63,11 +64,26 @@
      * to determine if a custom {@link CodecService} should be provided for the given index. A plugin that is not overriding
      * the {@link CodecService} through the plugin can ignore this method and the Codec specified in the {@link IndexSettings}
      * will be used.
+     *
+     * @deprecated Please use {@code getCustomCodecServiceFactory()} instead as it provides more context for {@link CodecService}
+     * instance construction.
      */
+    @Deprecated
     default Optional<CodecService> getCustomCodecService(IndexSettings indexSettings) {
         return Optional.empty();
     }
 
+    /**
+     * EXPERT:
+     * When an index is created this method is invoked for each engine plugin. Engine plugins can inspect the index settings
+     * to determine if a custom {@link CodecServiceFactory} should be provided for the given index. A plugin that is not overriding
+     * the {@link CodecServiceFactory} through the plugin can ignore this method and the default Codec specified in the
+     * {@link IndexSettings} will be used.
+     */
+    default Optional<CodecServiceFactory> getCustomCodecServiceFactory() {
```

Comment:
@dblock @andrross @nknize what do you think about this approach guys? this is non-breaking change but opens two ways to configure `CodecService` now (for 2.0.0 we could drop `getCustomCodecService` actually)

Reply:
Thanks @andrross ! Yes, super trivial: 

```
Optional<CodecService> getCustomCodecService(IndexSettings indexSettings) {
    return getCustomCodecServiceFactory().map(f -> f.createCodecService(new CodecServiceConfig(indexSettings, null));
}
```

The `CodecServiceConfig` also gives an opportunity to include more things in the future (if necessary).

Reply:
Why doesnt this take indexSettings as an argument? If I have an engine plugins implementing this method and want to create an unrelated index, wouldn't this getCustomCodecServiceFactory still be used?

Reply:
Could this be an issue: https://github.com/opensearch-project/OpenSearch/pull/2177/files#diff-5eecdf54afeae66f50a535c5d634cf34bc5811f67639189dbc13f3b65e7573d4R113? If index setting parsing is done after this check?

Reply:
@jmazanec15 it does not take the `indexSettings` because it just a factory, but the factory's method `createCodecService`, exactly where the codec service instance is expected to be created, takes the `indexSettings` as part of `CodecServiceConfig`. It also opens the door to add more services in the future without changing the method signatures.

Reply:
But wouldnt this scenario fail: I have EnginePlugin1 and EnginePlugin2 that both return non-null for getCustomCodecServiceFactory. There is an index setting, "engine_plugin_1" :"true/false" and "engine_plugin_2" :"true/false".

I create an index with  "engine_plugin_1" :"true" and "engine_plugin_2" :"false". Wouldnt this fail here: https://github.com/opensearch-project/OpenSearch/pull/2177/files#diff-5eecdf54afeae66f50a535c5d634cf34bc5811f67639189dbc13f3b65e7573d4R103?

Reply:
Ah I see what you mean, yeah, I think we could add `indexSettings` here so the plugin could decide it the factory should be provided or not.

Reply:
@jmazanec15 done, thanks for the feedback!